File: 3823-count-islands-with-total-value-divisible-by-k/3823-count-islands-with-total-value-divisible-by-k.c
int countIslands(int** grid, int gridSize, int* gridColSize, int k) {
    int n = gridSize, m = (*gridColSize);
    int cnt = 0;
    bool** vis = (bool**)malloc(sizeof(bool*) * n);
    
    for (int i = 0; i < n; i++)
        vis[i] = (bool*)calloc(m, sizeof(bool));

    long long dfs (int r, int c) {
        if (r >= n || r < 0 || c >= m || c < 0 || vis[r][c] || grid[r][c] == 0)
            return 0;

        vis[r][c] = true;
        long long sum = grid[r][c];
        sum += dfs (r - 1, c);
        sum += dfs (r, c + 1);
        sum += dfs (r + 1, c);
        sum += dfs (r, c - 1);
            
        return sum;
    }
    
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            if (!vis[i][j] && grid[i][j]) {
                long long val = dfs (i, j);
                if (val % k == 0)
                    cnt++;
            }
        }
    }
    return cnt;
}

File: 3823-count-islands-with-total-value-divisible-by-k/3823-count-islands-with-total-value-divisible-by-k.go
func countIslands(grid [][]int, k int) int {
    n, m := len(grid), len(grid[0])
    
    vis := make([][]bool, n)
    
    for i := 0; i < n; i++ {
        vis[i] = make([]bool, m)
    }
    var dfs func(r, c int) int64
    
    dfs = func(r, c int) int64 {
        if r >= n || r < 0 || c >= m || c < 0 || vis[r][c] || grid[r][c] == 0 {
            return 0
        }
        
        vis[r][c] = true
        var sum int64
        
        sum = int64(grid[r][c])
        
        sum += int64(dfs(r - 1, c))
        sum += int64(dfs(r, c + 1))
        sum += int64(dfs(r + 1, c))
        sum += int64(dfs(r, c - 1))
        
        return sum
    }

    cnt := 0
    for i := 0; i < n; i++ {
        for j := 0; j < m; j++ {
            if !vis[i][j] && grid[i][j] > 0 {
                val := dfs (i, j)
                if int(val) % k == 0 {
                    cnt++
                }
            }       
        }
    }
    return cnt
}